File: Algo-I/TPs/2-TP/Soluce/droite.py
import math

def is_equal(f1, f2, eps = 1e-6):
    return abs(f1 - f2) < eps

def droite(p1, p2):
    x1, y1 = p1
    x2, y2 = p2
    if is_equal(x1, x2) and is_equal(y1, y2):
        return None
    a, b = (y1 - y2, x2 - x1)
    c = a * x1 + b * y1
    return (a, b, c)

def appartient(d, p):
    a, b, c = d
    x, y = p
    return is_equal(a * x + b * y, c)

def paralleles(d1, d2):
    a1, b1, c1 = d1
    a2, b2, c2 = d2
    if is_equal(a1, 0) or is_equal(a2, 0):
        return a1 == a2
    return is_equal(float(b1) / a1, float(b2) / a2)

def intersection(d1, d2):
    a1, b1, c1 = d1
    a2, b2, c2 = d2
    det = a1 * b2 - b1 * a2
    if is_equal(det, 0):
        return None
    x = (c1 * b2 - b1 * c2) / det
    y = (a1 * c2 - c1 * a2) / det
    return (x, y)

def droite_normale(d, p):
    a, b, c = d
    x, y = p
    a1, b1 = (-b, a)
    c1 = a1 * x + b1 * y
    return (a1, b1, c1)

def symetrie_orthogonale(d, p):
    d_normale = droite_normale(d, p)
    xi, yi = intersection(d, d_normale)
    x, y = p
    return (2 * xi - x, 2 * yi - y)

def distance_droite_point(d, p):
    d_normale = droite_normale(d, p)
    xi, yi = intersection(d, d_normale)
    x, y = p
    return math.sqrt((x - xi)**2 + (y - yi)**2)

File: Algo-II/TPs/2 TP/Soluce/Droite.java
/**
 * Classe permettant de représenter une droite d'équation a*x+b*y+c=0.
 **/
public class Droite {
    /**
     * Précision utilisée dans les calculs.
     **/
    public static final double EPS = 1e-12;

    private double a;
    private double b;
    private double c;

    /**
     * Construit la droite d'equation a*x+b*y+c=0.
     * @param a Coefficient d'abscisse.
     * @param b Coefficient d'ordonnée.
     * @param c Terme indépendant.
     **/
    public Droite(double a, double b, double c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }

    /**
     * Retourne true si la droite est horizontale et false sinon.
     * @return true si horizontale, false sinon.
     **/
    public boolean isHorizontal() {
        return Math.abs(a) < EPS;
    }

    /**
     * Retourne l'intersection de la droite avec celle passée en paramètres ou
     * null si elles sont parallèles.
     * @param droite La droite avec laquelle calculer l'intersection.
     * @return l'intersection des droites ou null si elles sont parallèles.
     **/
    public Point intersection(Droite droite) {
        double det = a * droite.b - droite.a * b;
        //Les droites sont paralleles
        if (Math.abs(det) < EPS) {
            return null;
        }
        //Résolu par la méthode de Cramer. On prend l'opposé des résultats car
        //Cramer utilise l'équation a*x+b*y=c.
        double x = (droite.c * b - c * droite.b) / det;
        double y = (droite.a * c - a * droite.c) / det;
        return new Point(x, y);
    }

    /**
     * Retourne la droite passant par p1 et p2 et null si p1=p2.
     * @param p1 Le premier point.
     * @param p2 Le second point.
     * @return la droite passant par p1 et p2 et null si p1=p2.
     **/
    public static Droite create(Point p1, Point p2) {
        if (Math.abs(p1.getX() - p2.getX()) < EPS && Math.abs(p1.getY() - p2.getY()) < EPS) {
            return null;
        }
        double a = p1.getY() - p2.getY();
        double b = p2.getX() - p1.getX();
        double c = -(a * p1.getX() + b * p1.getY());
        return new Droite(a, b, c);
    }

    @Override
    public String toString() {
        return "" + a + "x+" + b + "y+" + c + "=0";
    }
}
